### PR TITLE
optimization for reading hive file  when all columns to read are partition keys

### PR DESCRIPTION
**Pull request by lgbo-ustc**
### Changelog category (leave one):
- Performance improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

When all the columns to read are partition keys, construct columns by the file's row number without real reading the hive file.

With following test sqls
```SQL
CREATE TABLE test_day_hour
(
    `rip` Nullable(String),
    `rtime` Nullable(String),
    `stattime` Nullable(Int64),
    `msgid` Nullable(String),
    `msgtype` Nullable(Int64),
    `day` String,
    `hour` String
)
ENGINE = Hive('thrift://xxxx', 'db', 'test_table')
PARTITION BY (day, hour);

SELECT
    day,
    count(*)
FROM test_day_hour
WHERE (day = '2022-05-01') AND (hour >= '20')
GROUP BY day

```

we have performance results.
* before optimize
```
1 row in set. Elapsed: 2.202 sec. Processed 99.18 million rows, 2.39 GB (45.04 million rows/s., 1.09 GB/s.)
```
* after optimize
```
1 row in set. Elapsed: 0.899 sec. Processed 99.18 million rows, 73.14 KB (110.33 million rows/s., 81.36 KB/s.)
```

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by lgbo-ustc** [comment]
> > Integration tests (release, actions) [1/2] — fail: 1
> 
> ```
> test_hive_query/test.py::test_cache_read_bytes
> ```

It is the case that all required colums are partition keys, and no data was read from file.  

Fixed it.

**Comment by lgbo-ustc** [comment]
The failures don't seem to be related to the hive part ?

File: src/Storages/Hive/StorageHive.cpp [hunk 1]
```diff
@@ -260,6 +273,45 @@ class StorageHiveSource : public SourceWithProgress, WithContext
...
+        for (size_t i = 0, sz = types.size(); i < sz; ++i)
+        {
+            if (!sample_block.has(names[i]))
+                continue;
+            auto col = types[i]->createColumnConst(rows, fields[i]);
+            auto col_idx = sample_block.getPositionByName(names[i]);
+            cols.insert(cols.begin() + col_idx, col);
+        }
+
+        if (source_info->need_file_column)
```
Comment (kssenii) [inline src/Storages/Hive/StorageHive.cpp:304]: what about path virtual column?

File: src/Storages/Hive/StorageHive.cpp [hunk 2]
```diff
@@ -151,13 +151,26 @@ class StorageHiveSource : public SourceWithProgress, WithContext
         {
             if (!reader)
             {
+                if (current_file_remained_rows) [[unlikely]]
```
Comment (kssenii) [inline src/Storages/Hive/StorageHive.cpp:154]: why unlikely? looks like it can happen if number of rows exceeds max block size

Comment (lgbo-ustc) [reply src/Storages/Hive/StorageHive.cpp:154]: I think that all the cols to read are partition keys maybe rare.

File: src/Storages/Hive/StorageHive.cpp (hunk 1 above)
Comment (lgbo-ustc) [reply src/Storages/Hive/StorageHive.cpp]: In this case, the path virtual col is not read. and I remove this code block

**Comment by kssenii** [comment]
> ClickHouse special build check (actions) — 6/7 artifact groups are OK

```
std=gnu++2a -MD -MT src/CMakeFiles/dbms.dir/Backups/BackupUtils.cpp.o -MF src/CMakeFiles/dbms.dir/Backups/BackupUtils.cpp.o.d -o src/CMakeFiles/dbms.dir/Backups/BackupUtils.cpp.o -c ../src/Backups/BackupUtils.cpp
May 17 02:25:15 /build/build_docker/../src/Backups/BackupUtils.cpp:369:22: error: 'auto &' can be declared as 'const auto &' [readability-qualified-auto,-warnings-as-errors]
May 17 02:25:15                 for (auto & [relative_path, backup_entry] : info.data)
May 17 02:25:15                      ^
May 17 02:25:15                      const 
May 17 02:25:15 /build/build_docker/../src/Backups/BackupUtils.cpp:377:18: error: 'auto &' can be declared as 'const auto &' [readability-qualified-auto,-warnings-as-errors]
May 17 02:25:15             for (auto & [relative_path, backup_entry] : info.data)
May 17 02:25:15                  ^
May 17 02:25:15                  const 
May 17 02:25:15 214897 warnings generated.
```

> Integration tests (release, actions) [2/2] — fail: 1

test_multiple_disks/test.py::test_concurrent_alter_move_and_drop[replicated]

**Review by kssenii — APPROVED** [review]

**Pull request merged**